File: templates/production-adapters/opencode-http-session/typescript/session-store.ts
import { join } from "node:path"

import { appendJsonLine, readJsonObject, writeJson } from "./json-store"

export type SessionRecord = {
  readonly agent: string
  readonly promptCount: number
  readonly sessionId: string
  readonly stage: string
}

type StorePayload = {
  runtimeState?: Record<string, unknown>
  sessions?: Record<string, { agent: string; id: string; promptCount: number }>
}

export class SessionStore {
  private readonly path: string
  private readonly turnsPath: string

  constructor(runDirectory: string) {
    this.path = join(runDirectory, "artifacts", "opencode-sessions.json")
    this.turnsPath = join(runDirectory, "artifacts", "opencode-turns.jsonl")
  }

  async get(stage: string): Promise<SessionRecord | undefined> {
    const payload = await this.load()
    const raw = payload.sessions?.[stage]
    if (raw === undefined) {
      return undefined
    }
    return { agent: raw.agent, promptCount: raw.promptCount, sessionId: raw.id, stage }
  }

  async upsert(stage: string, input: { readonly agent: string; readonly promptCount: number; readonly sessionId: string }): Promise<SessionRecord> {
    const payload = await this.load()
    payload.sessions ??= {}
    payload.sessions[stage] = { agent: input.agent, id: input.sessionId, promptCount: input.promptCount }
    await this.write(payload)
    return { agent: input.agent, promptCount: input.promptCount, sessionId: input.sessionId, stage }
  }

  async appendTurn(value: Record<string, unknown>): Promise<void> {
    await appendJsonLine(this.turnsPath, value)
  }

  async setRuntimeState(state: Record<string, unknown>): Promise<void> {
    const payload = await this.load()
    payload.runtimeState = state
    await this.write(payload)
  }

  async sessions(): Promise<Record<string, { agent: string; id: string; promptCount: number }>> {
    return (await this.load()).sessions ?? {}
  }

  private async load(): Promise<StorePayload> {
    if (!(await Bun.file(this.path).exists())) {
      return {}
    }
    return (await readJsonObject(this.path)) as StorePayload
  }

  private async write(payload: StorePayload): Promise<void> {
    await writeJson(this.path, payload)
  }
}

File: templates/production-adapters/opencode-http-session/python/session_store.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path


StateMap = dict[str, object]


@dataclass(frozen=True, slots=True)
class SessionRecord:
    stage: str
    session_id: str
    agent: str
    prompt_count: int


class SessionStore:
    def __init__(self, run_dir: Path) -> None:
        self._path = run_dir / "artifacts" / "opencode-sessions.json"
        self._turns_path = run_dir / "artifacts" / "opencode-turns.jsonl"

    def get(self, stage: str) -> SessionRecord | None:
        payload = self._load()
        sessions = payload.get("sessions")
        if not isinstance(sessions, dict):
            return None
        raw = sessions.get(stage)
        if not isinstance(raw, dict):
            return None
        session_id = raw.get("id")
        agent = raw.get("agent")
        prompt_count = raw.get("promptCount", 0)
        if not isinstance(session_id, str) or not isinstance(agent, str) or not isinstance(prompt_count, int):
            return None
        return SessionRecord(stage=stage, session_id=session_id, agent=agent, prompt_count=prompt_count)

    def upsert(self, stage: str, *, session_id: str, agent: str, prompt_count: int) -> SessionRecord:
        payload = self._load()
        sessions = payload.setdefault("sessions", {})
        if not isinstance(sessions, dict):
            raise RuntimeError("session store sessions payload is not a JSON object")
        sessions[stage] = {"id": session_id, "agent": agent, "promptCount": prompt_count}
        self._write(payload)
        return SessionRecord(stage=stage, session_id=session_id, agent=agent, prompt_count=prompt_count)

    def append_turn(self, record: dict[str, object]) -> None:
        self._turns_path.parent.mkdir(parents=True, exist_ok=True)
        with self._turns_path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(record, sort_keys=True) + "\n")

    def set_runtime_state(self, state: dict[str, object]) -> None:
        payload = self._load()
        payload["runtimeState"] = state
        self._write(payload)

    def _load(self) -> StateMap:
        if not self._path.is_file():
            return {}
        payload = json.loads(self._path.read_text(encoding="utf-8"))
        if not isinstance(payload, dict):
            raise RuntimeError("session store payload must be a JSON object")
        return dict(payload)

    def _write(self, payload: StateMap) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        self._path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")
